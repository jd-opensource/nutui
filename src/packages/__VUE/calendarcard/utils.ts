import Utils from '@/packages/utils/date'
import { CalendarCardDay, CalendarCardMonth, CalendarCardValue } from './types'

export const convertDateToDay = (date: Date) => {
  return date
    ? {
        year: date.getFullYear(),
        month: date.getMonth() + 1,
        date: date.getDate()
      }
    : null
}

export const convertDayToDate = (day: CalendarCardDay) => {
  return day ? new Date(day.year, day.month - 1, day.date) : null
}

export const valueToRange = (val?: CalendarCardValue) => {
  if (Array.isArray(val)) {
    return val.map((date: Date) => {
      return convertDateToDay(date)
    })
  }
  return val ? [convertDateToDay(val)] : []
}

export const rangeTovalue = (range?: CalendarCardDay[]) => {
  if (Array.isArray(range)) {
    return range.map((day: CalendarCardDay) => {
      return convertDayToDate(day)
    })
  }
  return range ? [convertDayToDate(range)] : []
}

/**
 * 获取当月面板中前一个月的日期数据
 */
export const getPrevMonthDays = (year: number, month: number, firstDayOfWeek: number) => {
  let prevMonth = month - 1
  let prevYear = year
  if (prevMonth <= 0) {
    prevMonth = 12
    prevYear -= 1
  }
  let days = Utils.getMonthPreDay(year, month)
  days -= firstDayOfWeek
  if (days >= 7) {
    days -= 7
  }

  const preDates = Utils.getMonthDays(`${prevYear}`, `${prevMonth}`)
  const months = Array.from(Array(preDates), (_, k) => {
    return {
      type: 'prev',
      year: prevYear,
      month: prevMonth,
      date: k + 1
    } as CalendarCardDay
  })
  return months.slice(preDates - days)
}

/**
 * 获取当前月的日期数据
 */
export const getCurrentMonthDays = (year: number, month: number) => {
  const days = Utils.getMonthDays(`${year}`, `${month}`)
  return Array.from(Array(days), (_, k) => {
    return {
      type: 'current',
      year,
      month,
      date: k + 1
    } as CalendarCardDay
  })
}

/**
 * 根据日期获取当前周的起始日期
 */
export const getCurrentWeekDays = (day: CalendarCardDay, firstDayOfWeek: number): CalendarCardDay[] => {
  const current = new Date(day.year, day.month - 1, day.date)
  const count = (current.getDay() + 7 - firstDayOfWeek) % 7
  return [
    convertDateToDay(new Date(current.getTime() - 24 * 60 * 60 * 1000 * count)) as CalendarCardDay,
    convertDateToDay(new Date(current.getTime() + 24 * 60 * 60 * 1000 * (6 - count))) as CalendarCardDay
  ]
}

/**
 * 获取 month 月份数据
 * @param month
 * @param firstDayOfWeek
 * @returns
 */
export const getDays = (month: CalendarCardMonth, firstDayOfWeek: number) => {
  const y = month.year
  const m = month.month
  const days = [...getPrevMonthDays(y, m, firstDayOfWeek), ...getCurrentMonthDays(y, m)] as CalendarCardDay[]
  const size = days.length
  const yearOfNextMonth = month.month === 12 ? month.year + 1 : month.year
  const monthOfNextMonth = month.month === 12 ? 1 : month.month + 1
  // 补全 6 行 7 列视图
  for (let i = 1; i <= 42 - size; i++) {
    days.push({
      type: 'next',
      year: yearOfNextMonth,
      month: monthOfNextMonth,
      date: i
    })
  }
  return days
}

/**
 * 比较两个日期
 * @param day1
 * @param day2
 * @returns
 */
export const compareDay = (day1: CalendarCardDay, day2: CalendarCardDay) => {
  if (day1 && day2) {
    if (day1.year === day2.year) {
      if (day1.month === day2.month) {
        return day1.date - day2.date
      }
      return day1.month - day2.month
    }
    return day1.year - day2.year
  }
}

export const isSameDay = (day1: CalendarCardDay, day2: CalendarCardDay) => {
  return compareDay(day1, day2) === 0
}
