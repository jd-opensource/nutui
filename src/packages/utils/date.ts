const Utils = {
  /**
   * 是否为闫年
   * @return {Boolse} true|false
   */
  isLeapYear: function (y: number): boolean {
    return (y % 4 == 0 && y % 100 != 0) || y % 400 == 0;
  },

  /**
   * 返回星期数
   * @return {String}
   */
  getWhatDay: function (year: number, month: number, day: number): string {
    const date = new Date(year + '/' + month + '/' + day);
    const index = date.getDay();
    const dayNames = ['星期日', '星期一', '星期二', '星期三', '星期四', '星期五', '星期六'];
    return dayNames[index];
  },

  /**
   * 返回星期数
   * @return {Number}
   */
  getMonthPreDay: function (year: number, month: number): number {
    const date = new Date(year + '/' + month + '/01');
    let day = date.getDay();
    if (day == 0) {
      day = 7;
    }
    return day;
  },

  /**
   * 返回月份天数
   * @return {Number}
   */
  getMonthDays: function (year: string, month: string): number {
    if (/^0/.test(month)) {
      month = month.split('')[1];
    }
    return ([0, 31, this.isLeapYear(Number(year)) ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31] as number[])[
      month as any
    ];
  },

  /**
   * 补齐数字位数
   * @return {string}
   */
  getNumTwoBit: function (n: number): string {
    n = Number(n);
    return (n > 9 ? '' : '0') + n;
  },

  /**
   * 日期对象转成字符串
   * @return {string}
   */
  date2Str: function (date: Date, split?: string): string {
    split = split || '-';
    const y = date.getFullYear();
    const m = this.getNumTwoBit(date.getMonth() + 1);
    const d = this.getNumTwoBit(date.getDate());
    return [y, m, d].join(split);
  },

  /**
   * 返回日期格式字符串
   * @param {Number} 0返回今天的日期、1返回明天的日期，2返回后天得日期，依次类推
   * @return {string} '2014-12-31'
   */
  getDay: function (i: number): string {
    i = i || 0;
    let date = new Date();
    const diff = i * (1000 * 60 * 60 * 24);
    date = new Date(date.getTime() + diff);
    return this.date2Str(date);
  },

  /**
   * 时间比较
   * @return {Boolean}
   */
  compareDate: function (date1: string, date2: string): boolean {
    const startTime = new Date(date1.replace('-', '/').replace('-', '/'));
    const endTime = new Date(date2.replace('-', '/').replace('-', '/'));
    if (startTime >= endTime) {
      return false;
    }
    return true;
  },

  /**
   * 时间是否相等
   * @return {Boolean}
   */
  isEqual: function (date1: string, date2: string): boolean {
    const startTime = new Date(date1).getTime();
    const endTime = new Date(date2).getTime();
    if (startTime == endTime) {
      return true;
    }
    return false;
  },
  getMonthWeek: function (year: string, month: string, date: string, firstDayOfWeek: number = 0): number {
    const dateNow = new Date(Number(year), parseInt(month) - 1, Number(date));
    let w = dateNow.getDay(); //星期数
    let d = dateNow.getDate();
    let remainder = 6 - w;
    if (firstDayOfWeek !== 0) {
      w = w == 0 ? 7 : w;
      remainder = 7 - w;
    }
    return Math.ceil((d + remainder) / 7);
  },
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  getYearWeek: function (year: string, month: string, date: string, firstDayOfWeek: number = 0): number {
    const dateNow = new Date(Number(year), parseInt(month) - 1, Number(date));
    const dateFirst = new Date(Number(year), 0, 1);
    const dataNumber = Math.round((dateNow.valueOf() - dateFirst.valueOf()) / 86400000);
    return Math.ceil((dataNumber + (dateFirst.getDay() + 1 - 1)) / 7);
  },
  getWeekDate: function (year: string, month: string, date: string, firstDayOfWeek: number = 0): string[] {
    const dateNow = new Date(Number(year), parseInt(month) - 1, Number(date));
    const nowTime = dateNow.getTime();
    let day = dateNow.getDay();
    if (firstDayOfWeek === 0) {
      const oneDayTime = 24 * 60 * 60 * 1000;
      //显示周日
      const SundayTime = nowTime - day * oneDayTime; //本周的周日
      //显示周六
      const SaturdayTime = nowTime + (6 - day) * oneDayTime; //本周的周六

      const sunday = this.date2Str(new Date(SundayTime));
      const saturday = this.date2Str(new Date(SaturdayTime));
      return [sunday, saturday];
    } else {
      day = day == 0 ? 7 : day;
      const oneDayTime = 24 * 60 * 60 * 1000;
      //显示周一
      const MondayTime = nowTime - (day - 1) * oneDayTime; //本周的周一
      //显示周日
      const SundayTime = nowTime + (7 - day) * oneDayTime; //本周的周日

      const monday = this.date2Str(new Date(MondayTime));
      const sunday = this.date2Str(new Date(SundayTime));
      return [monday, sunday];
    }
  },
  formatResultDate: function (date: string) {
    let days = [...date.split('-')];
    days[2] = Utils.getNumTwoBit(Number(days[2]));
    days[3] = `${days[0]}-${days[1]}-${days[2]}`;
    days[4] = Utils.getWhatDay(+days[0], +days[1], +days[2]);
    return days;
  }
};

export default Utils;
