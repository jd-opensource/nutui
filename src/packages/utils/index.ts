export * from './useRelation';
export * from './useLocale';
