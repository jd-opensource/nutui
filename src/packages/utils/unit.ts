import { ComponentPublicInstance, nextTick } from 'vue';
import { VueWrapper, DOMWrapper } from '@vue/test-utils';

function getTouch(el: Element | Window, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5
  };
}

export function trigger(
  wrapper: VueWrapper<ComponentPublicInstance<any, any, any>> | DOMWrapper<Element> | Element | Window,
  eventName: string,
  x = 0,
  y = 0,
  options: any = {}
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper;
  const touchList = options.touchList || [getTouch(el, x, y)];

  if (options.x || options.y) {
    touchList.push(getTouch(el, options.x, options.y));
  }

  const event = document.createEvent('CustomEvent');
  event.initCustomEvent(eventName, true, true, {});

  Object.assign(event, {
    clientX: x,
    clientY: y,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList
  });

  el.dispatchEvent(event);

  return nextTick();
}
export async function mockScrollTop(value: number) {
  Object.defineProperty(window, 'scrollTop', { value, writable: true });
  trigger(window, 'scroll');
  return nextTick();
}

export function sleep(delay = 0): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, delay);
  });
}
