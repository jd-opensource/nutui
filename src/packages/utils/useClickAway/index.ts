import { Ref, isRef, onMounted, onUnmounted } from 'vue';

type TargetType = Element | Ref<Element | undefined>;

export function useClickAway(onClickAway: () => void, target: TargetType | TargetType[], eventName = 'touchstart') {
  const handler = (event: Event) => {
    const targets = Array.isArray(target) ? target : [target];
    const inTarget = targets.some((item) => {
      const targetElement = isRef(item) ? item.value : item;
      return !targetElement || targetElement.contains(event.target as Node);
    });
    if (!inTarget) {
      onClickAway();
    }
  };

  onMounted(() => {
    window.addEventListener(eventName, handler, true);
  });

  onUnmounted(() => {
    window.removeEventListener(eventName, handler, true);
  });
}
