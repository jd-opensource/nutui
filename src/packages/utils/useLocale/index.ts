import locale from '@/packages/locale';
import { getPropByPath, isFunction } from '../util';

export const useLocale = (name = '') => {
  return (keyPath: string, ...args: unknown[]) => {
    name = name.toLocaleLowerCase();
    const languages = locale.languages();
    let text = keyPath;
    if (name && name.startsWith('nut')) {
      text = `${name.slice(3)}.${keyPath}`;
    }
    const res = getPropByPath(languages, text) || getPropByPath(languages, keyPath);
    return isFunction(res) ? res(...args) : res;
  };
};
