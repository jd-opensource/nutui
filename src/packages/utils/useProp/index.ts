import { computed, getCurrentInstance } from 'vue';
import type { ComponentPublicInstance, ComputedRef } from 'vue';

export const useProp = <T>(name: string): ComputedRef<T | undefined> => {
  const vm = getCurrentInstance();
  if (!vm) {
    throw new Error('useProp must be called within a setup function');
  }
  const proxy = vm.proxy as ComponentPublicInstance;
  const props = proxy?.$props as Record<string, T>;
  return computed(() => props?.[name]);
};
