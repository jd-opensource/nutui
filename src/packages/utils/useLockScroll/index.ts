import { body } from '@/packages/utils/util'
let count = 0

const CLSNAME = 'nut-overflow-hidden'

export const useLockScroll = (isLock: () => boolean) => {
  const lock = () => {
    if (isLock()) {
      try {
        !count && body.classList.add(CLSNAME)
        count++
      } catch (error) {
        console.warn('[NutUI] <useLockScroll>', error)
      }
    }
  }

  const unlock = () => {
    if (isLock() && count) {
      try {
        count--
        !count && body.classList.remove(CLSNAME)
      } catch (error) {
        console.warn('[NutUI] <unlock>', error)
      }
    }
  }

  return [lock, unlock]
}
