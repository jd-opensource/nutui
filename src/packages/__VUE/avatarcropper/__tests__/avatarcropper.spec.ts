import 'vitest-canvas-mock';
import { mount } from '@vue/test-utils';
import AvatarCropper from '../index.vue';
import { sleep, trigger, triggerDrag } from '@/packages/utils/unit';
import { h } from 'vue';

const mockFile = new File([new ArrayBuffer(10000)], 'test.jpg', {
  type: 'image/jpg'
});

test('layout default slot', () => {
  const wrapper = mount(AvatarCropper, {
    slots: {
      default: h('img', {
        src: 'https://img12.360buyimg.com/imagetools/jfs/t1/196430/38/8105/14329/60c806a4Ed506298a/e6de9fb7b8490f38.png'
      })
    }
  });

  expect(wrapper.html()).toMatchSnapshot();
  expect(wrapper.find('.nut-avatar-cropper').html()).toContain(
    '<img src="https://img12.360buyimg.com/imagetools/jfs/t1/196430/38/8105/14329/60c806a4Ed506298a/e6de9fb7b8490f38.png">'
  );
});

test('should render base cutAvatar and type', async () => {
  const wrapper = mount(AvatarCropper);
  const up_load = wrapper.find('.nut-avatar-cropper');
  expect(up_load.exists()).toBe(true);
  const up_load1 = wrapper.find('.nut-avatar-cropper__input');
  expect(up_load1.attributes().type).toBe('file');
});

test('AvatarCropper: Select the image to open the crop window', async () => {
  const wrapper = mount(AvatarCropper);
  const input: any = wrapper.find<HTMLInputElement>('.nut-avatar-cropper__input');
  expect(input.exists()).toBe(true);
  const smallFile = new File([new ArrayBuffer(100)], 'small.jpg');
  Object.defineProperty(input.element, 'files', {
    get: vi.fn().mockReturnValue([mockFile, smallFile])
  });
  expect(wrapper.find('.nut-cropper-popup').attributes()).toHaveProperty('style', 'display: none;');
  await input.trigger('change');
  await sleep();
  expect(wrapper.find('.nut-cropper-popup').attributes()).toHaveProperty('style', '');
  const canvas = wrapper.find('.nut-cropper-popup__canvas');
  expect(canvas.exists()).toBe(true);

  const track = wrapper.find('.nut-cropper-popup__highlight');

  trigger(track, 'touchstart', 0, 0, { x: 0, y: 0 });
  trigger(track, 'touchmove', 20, 20, { x: 40, y: 60 });
  trigger(track, 'touchend', 20, 100, { x: 40, y: 60 });

  triggerDrag(track, 50, 60);
  const toolbar = wrapper.findAll('.nut-cropper-popup__toolbar-item');
  expect(toolbar.length).toBe(4);

  const cancel = toolbar[0];
  cancel.trigger('click');
  expect(wrapper.emitted('cancel')).toBeTruthy();
  expect(input.element.value).toBe('');
  await sleep();
  expect(wrapper.find('.nut-cropper-popup').attributes()).toHaveProperty('style', 'display: none;');

  const reset = toolbar[1];
  reset.trigger('click');
  expect(wrapper.vm.angle).toBe(0);

  const rotate = toolbar[2];
  rotate.trigger('click');
  expect(wrapper.vm.angle).toBe(90);
  triggerDrag(track, 1000, 2000);
  rotate.trigger('click');
  expect(wrapper.vm.angle).toBe(180);
  rotate.trigger('click');
  rotate.trigger('click');
  expect(wrapper.vm.angle).toBe(0);

  const confirm = toolbar[3];
  confirm.trigger('click');
  expect(wrapper.emitted('confirm')).toBeTruthy();
});
