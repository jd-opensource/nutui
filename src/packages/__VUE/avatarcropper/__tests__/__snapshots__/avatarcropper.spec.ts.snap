// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`layout default slot 1`] = `
"<div class="nut-avatar-cropper" data-edit-text="编辑"><img src="https://img12.360buyimg.com/imagetools/jfs/t1/196430/38/8105/14329/60c806a4Ed506298a/e6de9fb7b8490f38.png"><input type="file" accept="image/*" class="nut-avatar-cropper__input"></div>
<div class="nut-cropper-popup" style="display: none;"><canvas class="nut-cropper-popup__canvas"></canvas>
  <div class="nut-cropper-popup__highlight">
    <div class="highlight" style="width: 0px; height: 0px;"></div>
  </div>
  <div class="nut-cropper-popup__toolbar bottom">
    <div class="flex-sb">
      <div class="nut-cropper-popup__toolbar-item">
        <view class="nut-button nut-button--danger nut-button--normal nut-button--round">
          <view class="nut-button__wrap">
            <!--v-if-->
            <!--v-if-->
            <view class="">取消</view>
          </view>
        </view>
      </div>
      <div class="nut-cropper-popup__toolbar-item"><svg class="nut-icon nut-icon-refresh2" style="color: rgb(255, 255, 255);" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1024 1024" role="presentation">
          <path d="M771.938 315.077h199.55L958.358 469.99 758.81 364.964c-13.128-7.877-18.38-23.63-10.502-36.759 2.625-7.877 13.128-13.128 23.63-13.128zm-535.63 393.846H44.636L57.764 554.01l191.672 105.026c13.128 7.877 18.38 23.63 10.502 36.759-5.25 7.877-15.753 13.128-23.63 13.128zM509.374 1024C257.313 1024 44.636 845.456 5.251 596.02 0 575.016 15.754 556.637 36.76 551.386c21.005-2.626 42.01 10.502 44.636 31.507 34.133 210.052 215.302 362.339 427.98 362.339 191.671 0 362.338-128.657 417.476-312.452 5.252-21.005 28.882-34.133 49.887-26.256 21.006 5.251 34.134 28.882 26.257 49.887C937.354 871.713 735.179 1024 509.375 1024zm467.364-551.385c-18.379 0-36.759-13.128-39.384-34.133C903.22 231.056 722.05 78.77 509.374 78.77c-191.671 0-362.338 128.657-414.85 312.452-5.252 21.005-28.883 34.133-49.888 26.256-21.005-5.251-34.133-28.882-26.257-49.887C81.395 152.287 283.57 0 509.374 0c252.062 0 464.739 178.544 504.123 427.98 2.626 21.005-10.502 42.01-31.507 44.635h-5.252z" fill="currentColor" fill-opacity="0.9"></path>
        </svg></div>
      <div class="nut-cropper-popup__toolbar-item"><svg class="nut-icon nut-icon-retweet" style="color: rgb(255, 255, 255);" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1024 1024" role="presentation">
          <path d="M136 552h63.6c4.4 0 8-3.6 8-8V288.7h528.6v72.6c0 1.9.6 3.7 1.8 5.2 2.9 3.6 8.1 4.3 11.7 1.4L893 255.4c4.3-5 3.6-10.3 0-13.2L749.7 129.8c-1.5-1.2-3.3-1.8-5.2-1.8-4.6 0-8.4 3.8-8.4 8.4V209H199.7c-39.5 0-71.7 32.2-71.7 71.8V544c0 4.4 3.6 8 8 8zm752-80h-63.6c-4.4 0-8 3.6-8 8v255.3H287.8v-72.6c0-1.9-.6-3.7-1.8-5.2-2.9-3.6-8.1-4.3-11.7-1.4L131 768.6c-4.3 5-3.6 10.3 0 13.2l143.3 112.4c1.5 1.2 3.3 1.8 5.2 1.8 4.6 0 8.4-3.8 8.4-8.4V815h536.6c39.5 0 71.7-32.2 71.7-71.8V480c-.2-4.4-3.8-8-8.2-8z" fill="currentColor" fill-opacity="0.9"></path>
        </svg></div>
      <div class="nut-cropper-popup__toolbar-item">
        <view class="nut-button nut-button--success nut-button--normal nut-button--round">
          <view class="nut-button__wrap">
            <!--v-if-->
            <!--v-if-->
            <view class="">确定</view>
          </view>
        </view>
      </div>
    </div>
  </div>
</div>"
`;
